h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoderModel.java

```diff
@@ -84,48 +85,79 @@ protected boolean defaultDropConsCols() {
+    public final ColumnsMapping[] _input_to_output_columns; // maps input columns (or groups of columns) to their corresponding encoded one(s).
     public final IcedHashMap<String, Frame> _target_encoding_map;
     public final IcedHashMap<String, Boolean> _te_column_to_hasNAs; //XXX: Map is a wrong choice for this, IcedHashSet would be perfect though
     
-    public TargetEncoderOutput(TargetEncoder b, IcedHashMap<String, Frame> teMap) {
-      super(b);
-      _parms = b._parms;
-      _nclasses = b.nclasses();
+    
+    public TargetEncoderOutput(TargetEncoder te) {
+      this(te, new IcedHashMap<>(), new ColumnsToSingleMapping[]{});
+    }
+    
+    public TargetEncoderOutput(TargetEncoder te, IcedHashMap<String, Frame> teMap, ColumnsToSingleMapping[] columnsToEncodeMapping) {
+      super(te);
+      _parms = te._parms;
+      _nclasses = te.nclasses();
       _target_encoding_map = teMap;
+      _input_to_encoding_column = columnsToEncodeMapping;
+      _input_to_output_columns = buildInOutColumnsMapping();
+      _te_column_to_hasNAs = buildCol2HasNAsMap();
       _model_summary = constructSummary();
+    }
 
-      _te_column_to_hasNAs = buildCol2HasNAsMap();
+    /**
+     * builds the name of encoded columns
+     */
+    private ColumnsMapping[] buildInOutColumnsMapping() {
+      ColumnsMapping[] encMapping = new ColumnsMapping[_input_to_encoding_column.length];
+      for (int i=0; i < encMapping.length; i++) {
+        ColumnsToSingleMapping toEncode = _input_to_encoding_column[i];
+        String[] groupCols = toEncode.from();
+        String columnToEncode = toEncode.toSingle();
+        Frame encodings = _target_encoding_map.get(columnToEncode);
+        String[] encodedColumns = listUsedTargetClasses().mapToObj(tc -> 
+                encodedColumnName(columnToEncode, tc, encodings.vec(TARGETCLASS_COL))
+        ).toArray(String[]::new);
+        encMapping[i] = new ColumnsMapping(groupCols, encodedColumns);
+      }
+      return encMapping;
     }
 
     private IcedHashMap<String, Boolean> buildCol2HasNAsMap() {
       final IcedHashMap<String, Boolean> col2HasNAs = new IcedHashMap<>();
       for (Map.Entry<String, Frame> entry : _target_encoding_map.entrySet()) {
         String teColumn = entry.getKey();
         Frame encodingsFrame = entry.getValue();
-        boolean hasNAs = _parms.train().vec(teColumn).cardinality() < encodingsFrame.vec(teColumn).cardinality(); //XXX: _parms.train().vec(teColumn).naCnt() > 0 ?
+        int teColCard = _parms.train().vec(teColumn) == null ? -1 : _parms.train().vec(teColumn).cardinality();
+        // if teColumn is a transient (generated for interactions), it can't have NAs as they're all encoded.
+        boolean hasNAs = teColCard > 0 && teColCard < encodingsFrame.vec(teColumn).cardinality(); //XXX: _parms.train().vec(teColumn).naCnt() > 0 ?
```

Comment:
Forgotten comment ?

Reply:
the comment was intended to clarify the line above: i'll inline it and rephrase it.